[logging] A few fixes/updates to record_compilation_metrics

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143332

Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Sort collections in `collection_to_str`
* Print non-string elements as `"<unknown>"` instead of None (since we don't expect non-strings)
* Move the population of the legacy metrics and any pre-processing to a new factory method in CompilationMetrics

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```
Internal testing: https://fburl.com/scuba/dynamo_compile/sandbox/l0me8auf

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

masnesral committed:
[logging] A few fixes/updates to record_compilation_metrics

Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Don't modify the input dict in `collection_to_str`
* Sort collections in `collection_to_str`
* Print non-string elements as "<unknown>" instead of None (since we don't expect non-strings)

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```

[ghstack-poisoned]

masnesral committed:
Update on "[logging] A few fixes/updates to record_compilation_metrics"


Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Don't modify the input dict in `collection_to_str`
* Sort collections in `collection_to_str`
* Print non-string elements as "<unknown>" instead of None (since we don't expect non-strings)

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral committed:
Update on "[logging] A few fixes/updates to record_compilation_metrics"


Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Don't modify the input dict in `collection_to_str`
* Sort collections in `collection_to_str`
* Print non-string elements as "<unknown>" instead of None (since we don't expect non-strings)

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1013,30 +1013,29 @@ def record_compilation_metrics(
 ):
-    def us_to_s(field):
-        metric = metrics.get(field, None)
+    def us_to_s(m: Dict[str, Any], field: str) -> Optional[float]:
```
Not necessarily needed, but I changed this to pass in the dict instead of assuming the `metrics` parameter of the outer function. Now that we have more than one dict this function operates on, it just seemed like a good way to make a future mistake.

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1013,30 +1013,29 @@ def record_compilation_metrics(
                 return str(item)
             except Exception:
-                return str(None)
+                return "<unknown>"
```
I could put these back to None if you prefer. I just thought it would be easier to spot real problems since many of our current scuba columns are often None.

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1013,30 +1013,29 @@ def record_compilation_metrics(
+        if not isinstance(metric, (set, list)):
+            return "<unknown>"
+
+        return ",".join(safe_str(item) for item in sorted(metric))
```
Note the "sorted". I don't know of any problems currently, just given the in-person discussion today I thought I may as well change it since I was already touching this code.

ppanchalia commented on torch/_dynamo/utils.py:
```diff
@@ -1088,23 +1074,48 @@ def safe_str(item: Any) -> str:
+    all_metrics["compile_id"] = str(compile_id) if compile_id else None
+
+    # Process a few of the fields before logging.
+    all_metrics["inductor_fx_remote_cache_hit_keys"] = collection_to_str(
+        all_metrics, "inductor_fx_remote_cache_hit_keys"
     )
+    all_metrics["inductor_fx_remote_cache_miss_keys"] = collection_to_str(
+        all_metrics, "inductor_fx_remote_cache_miss_keys"
+    )
```
I wonder if a better way to organize this would be to move `collection_to_str()` to CompilationMetrics and pass the list of fields to it and do the processing like

```
compilation_metrics = CompilationMetrics(**all_metrics)
compilation_metrics.process_collections(fields=["inductor_fx_remote_cache_miss_keys", "inductor_fx_remote_cache_hit_keys"])
```
I imagine there would be more such fields in the future that could use it

masnesral committed:
Update on "[logging] A few fixes/updates to record_compilation_metrics"


Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Don't modify the input dict in `collection_to_str`
* Sort collections in `collection_to_str`
* Print non-string elements as `"<unknown>"` instead of None (since we don't expect non-strings)

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```
Internal testing: https://fburl.com/scuba/dynamo_compile/sandbox/sapekgfq

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ppanchalia commented on torch/_dynamo/utils.py:
```diff
@@ -1061,54 +1105,32 @@ def safe_str(item: Any) -> str:
+    compile_id = metrics.get("compile_id")
+    if not compile_id:
+        compile_id = torch._guards.CompileContext.current_compile_id()
+    all_metrics["compile_id"] = str(compile_id) if compile_id else None
```
Can we add a set_compile_id() method to CompilationMetrics, the new CompilationMetrics is so beautiful with the factory method and all, this adhoc key assignment sticks out.

masnesral replied:
Ok, I know how to fix this

masnesral committed:
Update on "[logging] A few fixes/updates to record_compilation_metrics"


Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Sort collections in `collection_to_str`
* Print non-string elements as `"<unknown>"` instead of None (since we don't expect non-strings)
* Move the population of the legacy metrics and any pre-processing to a new factory method in CompilationMetrics

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```
Internal testing: https://fburl.com/scuba/dynamo_compile/sandbox/l0me8auf

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral committed:
Update on "[logging] A few fixes/updates to record_compilation_metrics"


Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Sort collections in `collection_to_str`
* Print non-string elements as `"<unknown>"` instead of None (since we don't expect non-strings)
* Move the population of the legacy metrics and any pre-processing to a new factory method in CompilationMetrics

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```
Internal testing: https://fburl.com/scuba/dynamo_compile/sandbox/l0me8auf

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge -i

malfet commented:
@pytorchbot revert -m "Surprisingly failure is caused by this PR" -c ignoredsignal

malfet commented:
Actually, it breaks the reraise test across the board (though all one needs to do is just to adjust the test:
> FAILED [0.0491s] dynamo/test_exceptions.py::ExceptionTests::test_nn_reraise - AssertionError: String comparison failed:

masnesral committed:
Update on "[logging] A few fixes/updates to record_compilation_metrics"


Summary: Mostly cosmetic, but one bug fix:
* Bug fix: Make sure compile_id is converted to a string in the compilation metrics so it's printed as, e.g., "0/1" instead of "[0, 1]"
* Sort collections in `collection_to_str`
* Print non-string elements as `"<unknown>"` instead of None (since we don't expect non-strings)
* Move the population of the legacy metrics and any pre-processing to a new factory method in CompilationMetrics

Test Plan:
```
python test/dynamo/test_structured_trace.py
python test/dynamo/test_utils.py
```
Internal testing: https://fburl.com/scuba/dynamo_compile/sandbox/l0me8auf

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
